presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -0,0 +1,227 @@
+            return Optional.empty();
+        }
+
+        if (exchangeNode.getType() == REPARTITION) {
+            // if partitioning columns are not a subset of grouping keys,
+            // we can't push this through
+            List<Symbol> partitioningColumns = exchangeNode.getPartitioningScheme()
+                    .getPartitioning()
+                    .getArguments()
+                    .stream()
+                    .filter(Partitioning.ArgumentBinding::isVariable)
+                    .map(Partitioning.ArgumentBinding::getColumn)
+                    .collect(Collectors.toList());
+
+            if (!aggregationNode.getGroupingKeys().containsAll(partitioningColumns)) {
+                return Optional.empty();
+            }
+        }
+
+        // currently, we only support plans that don't use pre-computed hash functions
+        if (aggregationNode.getHashSymbol().isPresent() || exchangeNode.getPartitioningScheme().getHashColumn().isPresent()) {
+            return Optional.empty();
+        }
+
+        switch (aggregationNode.getStep()) {
+            case SINGLE:
+                // Split it into a FINAL on top of a PARTIAL and
+                // reprocess the resulting plan to push the partial
+                // below the exchange (see case below).
+                return Optional.of(split(aggregationNode, context));
+            case PARTIAL:
+                // Push it underneath each branch of the exchange
+                // and reprocess in case it can be pushed further down
+                // (e.g., if there are local/remote exchanges stacked)
+                return Optional.of(pushPartial(aggregationNode, exchangeNode, context));
+            default:
+                return Optional.empty();
+        }
+    }
+
+    private PlanNode pushPartial(AggregationNode partial, ExchangeNode exchange, Context context)
+    {
+        List<PlanNode> partials = new ArrayList<>();
+        for (int i = 0; i < exchange.getSources().size(); i++) {
+            PlanNode source = exchange.getSources().get(i);
+
+            SymbolMapper.Builder mappingsBuilder = SymbolMapper.builder();
+            for (int outputIndex = 0; outputIndex < exchange.getOutputSymbols().size(); outputIndex++) {
+                Symbol output = exchange.getOutputSymbols().get(outputIndex);
+                Symbol input = exchange.getInputs().get(i).get(outputIndex);
+                if (!output.equals(input)) {
+                    mappingsBuilder.put(output, input);
+                }
+            }
+
+            SymbolMapper symbolMapper = mappingsBuilder.build();
+            AggregationNode mappedPartial = symbolMapper.map(partial, source, context.getIdAllocator());
+
+            Assignments.Builder assignments = Assignments.builder();
+
+            for (Symbol output : partial.getOutputSymbols()) {
+                Symbol input = symbolMapper.map(output);
+                assignments.put(output, input.toSymbolReference());
+            }
```

Comment:
Since you are project-off columns I think that such code would be more clear:

```
Assignments projectOffAssignments = Assignments.identity(partial.getOutputSymbols().stream()
     .map(output -> symbolMapper.map(ouptut).toSymbolReference()
     .colect(toList());
```